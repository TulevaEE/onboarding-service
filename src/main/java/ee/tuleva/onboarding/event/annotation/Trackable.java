package ee.tuleva.onboarding.event.annotation;

import static java.lang.annotation.ElementType.METHOD;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

import ee.tuleva.onboarding.event.TrackableEventType;
import java.lang.annotation.Retention;
import java.lang.annotation.Target;

@Target(METHOD)
@Retention(RUNTIME)
public @interface Trackable {
  TrackableEventType value();
}
