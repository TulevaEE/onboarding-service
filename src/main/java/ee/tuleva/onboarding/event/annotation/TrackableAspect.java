package ee.tuleva.onboarding.event.annotation;

import static java.util.stream.Collectors.toMap;

import com.fasterxml.jackson.databind.JavaType;
import com.fasterxml.jackson.databind.ObjectMapper;
import ee.tuleva.onboarding.auth.principal.Person;
import ee.tuleva.onboarding.event.TrackableEventPublisher;
import java.util.Map;
import java.util.stream.IntStream;
import lombok.RequiredArgsConstructor;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.stereotype.Component;

@Aspect
@Component
@RequiredArgsConstructor
public class TrackableAspect {

  private final TrackableEventPublisher trackableEventPublisher;

  private final ObjectMapper objectMapper;

  @Before("@annotation(trackable) && args(.., person)")
  public void track(JoinPoint joinPoint, Trackable trackable, Person person) {
    String[] argNames = ((MethodSignature) joinPoint.getSignature()).getParameterNames();
    Object[] args = joinPoint.getArgs();

    Map<String, Object> params =
        IntStream.range(0, args.length - 1).boxed().collect(toMap(i -> argNames[i], i -> args[i]));

    JavaType type =
        objectMapper.getTypeFactory().constructMapType(Map.class, String.class, Object.class);
    Map<String, Object> data = objectMapper.convertValue(params, type);
    trackableEventPublisher.publish(person, trackable.value(), data);
  }
}
