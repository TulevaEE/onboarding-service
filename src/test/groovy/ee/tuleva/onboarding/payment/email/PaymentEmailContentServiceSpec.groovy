package ee.tuleva.onboarding.payment.email

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.annotations.SnapshotName
import au.com.origin.snapshots.spock.EnableSnapshots
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import spock.lang.Specification

import static ee.tuleva.onboarding.auth.UserFixture.sampleUser
import static ee.tuleva.onboarding.epis.contact.ContactDetailsFixture.contactDetailsFixture
import static ee.tuleva.onboarding.payment.PaymentFixture.aNewPayment

@SpringBootTest
@EnableSnapshots
class PaymentEmailContentServiceSpec extends Specification {

  @Autowired
  PaymentEmailContentService emailContentService

  private Expect expect

  @SnapshotName("third_pillar_payment_success")
  def "#language: renders the third pillar payment success email correctly"() {
    given:
    def user = sampleUser().build()
    def payment = aNewPayment()
    def contactDetails = contactDetailsFixture()
    def locale = Locale.forLanguageTag(language)

    when:
    String html =
        emailContentService.getThirdPillarPaymentSuccessHtml(user, payment, contactDetails, locale)

    then:
    expect.scenario(language).toMatchSnapshot(html)

    where:
    language | _
    'en'     | _
    'et'     | _
  }

  @SnapshotName("third_pillar_suggest_second")
  def "#language: renders the third pillar suggest second email correctly"() {
    given:
    def user = sampleUser().build()
    def locale = Locale.forLanguageTag(language)

    when:
    String html = emailContentService.getThirdPillarSuggestSecondHtml(user, locale)

    then:
    expect.scenario(language).toMatchSnapshot(html)

    where:
    language | _
    'en'     | _
    'et'     | _
  }
}
