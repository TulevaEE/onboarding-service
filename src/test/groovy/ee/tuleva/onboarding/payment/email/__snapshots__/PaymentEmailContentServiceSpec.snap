third_pillar_payment_success[en]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Welcome Jordan,</p>
    <p>I’m glad you’ve decided to start saving in the third pillar with us.</p>
    <p>Your payment has successfully arrived at the Estonian Pension Registry. Your new contribution will appear on your pension account by the end of the next working day.</p>

    <p>If you have any questions please contact us via <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a> or 644 5100.</p>
    <br/>
    <p>Tõnu Pekk<br>Fund manager<br>Tuleva</p>
  </body>
</html>

]


third_pillar_payment_success[et]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Tere tulemast, Jordan.</p>
    <p>Mul on hea meel, et oled alustanud meiega koos kolmandasse sambasse raha kogumist.</p>
    <p>Sinu sissemakse on edukalt Pensionikeskusesse jõudnud. Uued osakud jõuavad sinu pensionikontole järgmise tööpäeva lõpuks.</p>

    <p>Kui sul tekib ükskõik milliseid küsimusi, võta kõhklemata ühendust: e-post <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a>, tel 644 5100.</p>
    <br/>
    <p>Tõnu Pekk<br>Fondijuht<br>Tuleva</p>
  </body>
</html>

]


third_pillar_suggest_second[en]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Hello, Jordan.</p>
    <p>You have started to save in Tuleva’s third pillar fund with low fees and a fact-based investment strategy. I noticed that you are not investing your second pillar with us.<br><br>In a typical bank-run fund you pay up to three times more in fees than with Tuleva. You can check where your second pillar contributions go by <a href="https://pension.tuleva.ee/login?utm_source=autoemail&utm_medium=clients&utm_campaign=sug_2.1">logging in here</a>. Switching the fund is free of charge and takes only a few minutes.</p>

    <br/>
    <strong>Should you bring your second pillar to Tuleva?</strong>
    <p>I would do that – actually, I have my second pillar in Tuleva World Stocks Pension Fund and my third pillar also with Tuleva.<br><br>In fact, banks also have low-cost index funds in their product offering, but only with Tuleva can you be sure that the good funds are not hidden among high-fee ones that have so far performed pretty badly. Nobody at Tuleva will try to convince you to switch back to a high-cost fund.</p>

    <br/>
    <strong>Isn’t it better for risk diversification to keep my second pillar elsewhere?</strong>
    <p>All Estonian pension funds – both in the second and third pillar – have their assets well diversified among hundreds of investments. In addition, in all pension funds the investors are protected against the business risks of the fund manager. This is set in the Law and investors in pension funds themselves do not have to spread their assets between many fund managers for risk protection.<br><br>It is important to remember, however, that your money in pension funds is open to market risk. When world markets fall, all pension fund unit prices fall too. It does not help if your assets are spread among many fund managers.</p>

    <br/>
    <strong>Do you want to know why the second and third pillars are useful?</strong>
    <p><a href="https://tuleva.ee/en/recommendations/15-minutes/?utm_source=autoemail&utm_medium=clients&utm_campaign=sug_2.1&utm_content=laura_rikkaks">Read this short article</a> where I describe my investment portfolio, the foundation of which are the second and third pension pillars. You will also find links to a longer series of blog posts which should give you the confidence to start investing now. You will find that pension funds are very clever investment tools with significant tax advantages.</p>
    <br/>

    <p>If you have any questions please contact us via <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a> or 644 5100.</p>
    <br/>

    <p>Tõnu Pekk<br>Fund manager<br>Tuleva</p>
  </body>
</html>

]


third_pillar_suggest_second[et]=[
<!DOCTYPE HTML>
<html>
  <body>
    <p>Tervist, Jordan.</p>
    <p>Sa alustasid kogumist Tuleva madalate kulude ja tõestuspõhise investeerimisstrateegiaga kolmanda samba fondis. Märkasin, et teise sambasse sa veel koos meiega ei kogu.<br><br>Vanas pangafondis maksad igal aastal oma varast tasudeks kuni kolm korda rohkem. Vaata üle, kuhu sinu teise samba säästud täna lähevad - selleks <a href="https://pension.tuleva.ee/login?utm_source=autoemail&utm_medium=clients&utm_campaign=sug_2.1">logi sisse oma pensionikontole</a>. Fondivahetus on tasuta ja võtab vaid mõne minuti.</p>

    <br/>
    <strong>Kas sa peaksid oma teise samba just Tulevasse tooma?</strong>
    <p>Mina sinu asemel tooksin küll - ma ise kogungi teise sambasse Tuleva Maailma Aktsiate Pensionifondis ja kolmandasse sambasse samuti Tulevas.<br><br>Tegelikult on täna ka pankadel soodsama tasuga indeksifondid valikus olemas, aga ainult Tulevas võid kindel olla, et head fondid pole peidetud kõrge tasu ja siiani järjekindlalt kehvema tootlusega fondide vahele. Keegi Tulevas ei hakka sind kallimasse fondi tagasi meelitama.</p>

    <br/>
    <strong>Kas riskide hajutamiseks pole parem koguda teise sambasse mujal?</strong>
    <p>Kõigi Eesti pensionifondide - nii teise kui kolmanda samba fondide - vara on maailmaturu sadade instrumentide vahel väga hästi hajutatud. Lisaks on pensionikoguja kõigis pensionifondides kaitstud fondivalitseja äririski eest. Seda tagab seadus ja pensionikoguja ise ei pea riski maandamiseks raha mitme fondivalitseja vahel laiali jagama.<br><br>Oluline on ka meeles pidada, et sinu raha kõigis pensionifondides on avatud tururiskile. Kui maailma väärtpaberiturud langevad, siis langevad kõigi pensionifondide osakute hinnad. Siin vara jagamine erinevate fondivalitsejate aktsiatesse investeerivate fondide vahel ei aita.</p>

    <br/>
    <strong>Tahaksid lähemalt teada, milleks teine ja kolmas sammas kasulikud on?</strong>
    <p><a href="https://tuleva.ee/laura-rikkaks/15-minutiga-miljonariks/?utm_source=autoemail&utm_medium=clients&utm_campaign=sug_2.1&utm_content=laura_rikkaks">Loe läbi lühike artikkel</a>, kus kirjeldan oma lihtsat investeerimisportfelli, mille aluseks on teine ja kolmas pensionisammas. Artikli alt leiad viited pikemale artikliseeriale, mis peaks andma sulle enesekindluse investeerimisega alustamiseks. Saad väga hea arusaamise sellest, miks pensionisambad on kavalad maksueelisega tööriistad tuleviku heaks raha kogumiseks.</p>
    <br/>

    <p>Kui sul tekib ükskõik milliseid küsimusi, võta kõhklemata ühendust: e-post <a href="mailto:tuleva@tuleva.ee">tuleva@tuleva.ee</a>, tel 644 5100.</p>
    <br/>

    <p>Tõnu Pekk<br>Fondijuht<br>Tuleva</p>
  </body>
</html>

]
